pkg/operation/common/utils.go

```diff
@@ -547,3 +550,107 @@ func CurrentReplicaCount(client client.Client, namespace, deploymentName string)
 	}
 	return *deployment.Spec.Replicas, nil
 }
+
+// RespectShootSyncPeriodOverwrite checks whether to respect the sync period overwrite of a Shoot or not.
+func RespectShootSyncPeriodOverwrite(respectSyncPeriodOverwrite bool, shoot *gardenv1beta1.Shoot) bool {
+	return respectSyncPeriodOverwrite || shoot.Namespace == GardenNamespace
+}
+
+// ShouldIgnoreShoot determines whether a Shoot should be ignored or not.
+func ShouldIgnoreShoot(respectSyncPeriodOverwrite bool, shoot *gardenv1beta1.Shoot) bool {
+	if !RespectShootSyncPeriodOverwrite(respectSyncPeriodOverwrite, shoot) {
+		return false
+	}
+
+	value, ok := shoot.Annotations[ShootIgnore]
+	if !ok {
+		return false
+	}
+
+	ignore, _ := strconv.ParseBool(value)
+	return ignore
+}
+
+// IsShootFailed checks if a Shoot is failed.
+func IsShootFailed(shoot *gardenv1beta1.Shoot) bool {
+	lastOperation := shoot.Status.LastOperation
+
+	return lastOperation != nil && lastOperation.State == gardencorev1alpha1.LastOperationStateFailed &&
+		shoot.Generation == shoot.Status.ObservedGeneration &&
+		shoot.Status.Gardener.Version == version.Get().GitVersion
+}
+
+// IsNowInEffectiveShootMaintenanceTimeWindow checks if the current time is in the effective
+// maintenance time window of the Shoot.
+func IsNowInEffectiveShootMaintenanceTimeWindow(shoot *gardenv1beta1.Shoot) bool {
+	return EffectiveShootMaintenanceTimeWindow(shoot).Contains(time.Now())
+}
+
+// IsOperationOngoing checks whether the Shoot's generation has changed or if the LastOperation status
+// is not Succeeded.
+func IsOperationOngoing(shoot *gardenv1beta1.Shoot) bool {
```

Comment:
I would maybe rename this function to "NeedsReconciliation" or "IsUpToDate" or something similar if you have even better proposals.